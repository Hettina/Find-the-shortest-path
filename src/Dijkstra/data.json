{
  "data" : [
    {"nodeName": "A","adjacentNodes": [{"nodeName": "C", "distance": 2}]},
    {"nodeName": "B", "adjacentNodes": [{"nodeName": "D", "distance": 4}, {"nodeName": "E", "distance": 7}]},
    {"nodeName": "C", "adjacentNodes" :[{"nodeName":"A", "distance":2}, {"nodeName":"D", "distance": 1}, {"nodeName":"F", "distance": 4}]},
    {"nodeName":"D", "adjacentNodes" :[{"nodeName":"B", "distance":4}, {"nodeName":"C", "distance":1}, {"nodeName":"F", "distance":1}, {"nodeName":"G", "distance":2}]},
    {"nodeName":"E", "adjacentNodes" :[{"nodeName":"B","distance": 7}, {"nodeName":"H","distance": 10}]},
    {"nodeName":"F", "adjacentNodes": [{"nodeName":"C","distance": 4}, {"nodeName":"D","distance": 1}, {"nodeName":"G","distance": 3}]},
    {"nodeName":"G", "adjacentNodes":[{"nodeName":"D", "distance": 2},{"nodeName":"F", "distance": 3},{"nodeName":"H", "distance": 4}]},
    {"nodeName":"H", "adjacentNodes":[{"nodeName":"E", "distance": 10},  {"nodeName":"G","distance": 4}]}
  ]
}

