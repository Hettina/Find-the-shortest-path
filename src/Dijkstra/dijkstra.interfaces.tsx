export interface DijkstraProps {
    start: string,
    end: string
}

export interface NodeObject {
    nodeName: string
    adjacentNodes: Array<AdjNodesObject>
}

export interface AdjNodesObject {
    nodeName: string,
    distance: number
}

export interface TableObject {
    nodeName: string,
    totalShortestDistance: number,
    previousNode: string
}
