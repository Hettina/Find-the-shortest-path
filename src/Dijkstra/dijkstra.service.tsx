import {AdjNodesObject, NodeObject, TableObject} from './dijkstra.interfaces'

export const getInitialTableValue = (start: string, nodes: Array<NodeObject>): Array<TableObject> => {
    return nodes.map( (node:NodeObject) => {
        const distance = start === node.nodeName ? 0 : Infinity;
        return {
            nodeName: node.nodeName,
            totalShortestDistance: distance,
            previousNode: ''
        }
    });
};

export const getCurrentNode = (nodes: Array<NodeObject>, currentNodeName: string) => nodes.filter((node: NodeObject) => node.nodeName === currentNodeName)[0];

export const getUniqueNodesWithTotalDistance = (adjacentNodes : Array<AdjNodesObject>, visitedArray: Array<string>, shortestDistance: number) => {
    let uniqueAdjacentNodes: any =[];

    adjacentNodes.forEach((adjacentNode: AdjNodesObject) => {
        if(!visitedArray.includes(adjacentNode.nodeName)) {
            uniqueAdjacentNodes.push({
                nodeName: adjacentNode.nodeName,
                distance: shortestDistance + adjacentNode.distance
            });
        }
    });
    return uniqueAdjacentNodes;
};

export const updateTable = (table: Array<TableObject>, nodesWithTotalDistance: Array<AdjNodesObject>, currentNodeName: string ) => {
    let newTable = table;
    nodesWithTotalDistance.forEach((element:AdjNodesObject) => {
        let row = newTable.filter((row:any) => row.nodeName === element.nodeName)[0];
        if (row && (row.totalShortestDistance ===Infinity || element.distance < row.totalShortestDistance )) {
            row.totalShortestDistance = element.distance;
            row.previousNode = currentNodeName;
        }
    });
    return newTable;
};

export const getNextNode= (table:Array<TableObject>, currentNodeName: string, visitedArray: Array<string>) => {
    let tempTable = table.filter((row:any) => {
        return (!visitedArray.includes(row.nodeName) && row.nodeName!==currentNodeName) && !(row.totalShortestDistance === 0 || row.totalShortestDistance === Infinity)
    });
    tempTable.sort((a:any, b:any) => a.totalShortestDistance - b.totalShortestDistance);

    return tempTable[0];
};

