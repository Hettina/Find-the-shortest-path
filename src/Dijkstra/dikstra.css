
.resultParagraph{
    font-size: 1.3em;
}

.answer {
    color: hotpink;
    font-weight: bolder;
}