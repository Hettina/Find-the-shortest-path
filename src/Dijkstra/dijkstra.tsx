import React from 'react';
import {
    getInitialTableValue,
    getCurrentNode,
    getUniqueNodesWithTotalDistance,
    updateTable,
    getNextNode
} from './dijkstra.service'
import './dikstra.css';

import {NodeObject, DijkstraProps} from './dijkstra.interfaces'
let nodes = require( './data.json');

const updateUnvisitedArray = (unvisitedArray: Array<string>, currentNodeName:string) => unvisitedArray.filter(item => item !== currentNodeName);

const Dijkstra: React.FC <DijkstraProps> = (props) => {

    const findShortestPath = (start: string, end: string, nodes: Array<NodeObject>): number => {
        if(start!=='' && end!== '') {
            // setup Initial
            let currentNodeName = start;
            let totalShortestDistance = 0;
            let visitedArray = [];
            let unvisitedArray = nodes.map((node: NodeObject) => node.nodeName);
            let table = getInitialTableValue(start, nodes);

            //loop through all the nodes
            if(unvisitedArray.includes(start)) {
                while (unvisitedArray.length > 0) {
                    let adjacentNodes = getCurrentNode(nodes, currentNodeName).adjacentNodes;
                    let nodesWithTotalDistance = getUniqueNodesWithTotalDistance(adjacentNodes, visitedArray, totalShortestDistance);
                    table = updateTable(table, nodesWithTotalDistance, currentNodeName);


                    // find next node to move to
                    let nextNode = getNextNode(table, currentNodeName, visitedArray);
                    if (nextNode) {
                        totalShortestDistance = nextNode.totalShortestDistance;
                        visitedArray.push(currentNodeName);
                        unvisitedArray = updateUnvisitedArray(unvisitedArray, currentNodeName);
                        currentNodeName = nextNode.nodeName;
                    } else {
                        unvisitedArray = updateUnvisitedArray(unvisitedArray, currentNodeName);
                    }
                }


                let endNode = table.find(row => row.nodeName === end);
                return endNode ? endNode.totalShortestDistance : 0;
            }
        }
            return 0;
    };

    return (
        <p className="resultParagraph">
            <span> The shortest path is: </span>
            <span className="answer">{findShortestPath(props.start, props.end, nodes.data)}</span>
        </p>
    );
};


export default Dijkstra;
