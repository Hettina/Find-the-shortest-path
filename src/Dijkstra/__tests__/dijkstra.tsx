import React from 'react';
import { render, cleanup } from 'react-testing-library';
import Dijkstra from '../dijkstra';

afterEach(cleanup);

describe('Test Dijkstra component', () => {
    it('renders when empty string is passed in fot start and end', () => {
        const container = render(<Dijkstra start={''} end={''}/>);
        expect(container).toMatchSnapshot()
    });
    it('renders when empty string is passed in for just the end', () => {
        const container = render(<Dijkstra start={'A'} end={''}/>);
        expect(container).toMatchSnapshot()
    });
    it('renders when values are passed in for start and end', () => {
        const container = render(<Dijkstra start={'A'} end={'C'}/>);
        expect(container).toMatchSnapshot()
    });
});

