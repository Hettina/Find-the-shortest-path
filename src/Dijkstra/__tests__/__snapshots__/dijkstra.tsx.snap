// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Dijkstra component renders when empty string is passed in for just the end 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="resultParagraph"
      >
        <span>
           The shortest path is: 
        </span>
        <span
          class="answer"
        >
          0
        </span>
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      class="resultParagraph"
    >
      <span>
         The shortest path is: 
      </span>
      <span
        class="answer"
      >
        0
      </span>
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Dijkstra component renders when empty string is passed in fot start and end 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="resultParagraph"
      >
        <span>
           The shortest path is: 
        </span>
        <span
          class="answer"
        >
          0
        </span>
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      class="resultParagraph"
    >
      <span>
         The shortest path is: 
      </span>
      <span
        class="answer"
      >
        0
      </span>
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Dijkstra component renders when values are passed in for start and end 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="resultParagraph"
      >
        <span>
           The shortest path is: 
        </span>
        <span
          class="answer"
        >
          2
        </span>
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      class="resultParagraph"
    >
      <span>
         The shortest path is: 
      </span>
      <span
        class="answer"
      >
        2
      </span>
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
